storage/database/src/config/config.rs

```diff
@@ -0,0 +1,210 @@
+//! The main [`Config`] struct, holding all configurable values.
+
+//---------------------------------------------------------------------------------------------------- Import
+use std::{borrow::Cow, num::NonZeroUsize, path::Path};
+
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+use crate::{config::SyncMode, constants::DATABASE_DATA_FILENAME, resize::ResizeAlgorithm};
+
+//---------------------------------------------------------------------------------------------------- Constants
+/// Default value for [`Config::reader_threads`].
+///
+/// ```rust
+/// use cuprate_database::config::*;
+/// assert_eq!(READER_THREADS_DEFAULT.get(), 126);
+/// ```
+pub const READER_THREADS_DEFAULT: NonZeroUsize = match NonZeroUsize::new(126) {
+    Some(n) => n,
+    None => unreachable!(),
+};
+
+//---------------------------------------------------------------------------------------------------- ConfigBuilder
+/// Builder for [`Config`].
+///
+// SOMEDAY: there's are many more options to add in the future.
+#[derive(Debug, Clone, PartialEq, PartialOrd)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct ConfigBuilder {
+    /// [`Config::db_directory`].
+    db_directory: Cow<'static, Path>,
+
+    /// [`Config::sync_mode`].
+    sync_mode: Option<SyncMode>,
+
+    /// [`Config::reader_threads`].
+    reader_threads: Option<NonZeroUsize>,
+
+    /// [`Config::resize_algorithm`].
+    resize_algorithm: Option<ResizeAlgorithm>,
+}
+
+impl ConfigBuilder {
+    /// Create a new [`ConfigBuilder`].
+    ///
+    /// [`ConfigBuilder::build`] can be called immediately
+    /// after this function to use default values.
+    pub const fn new(db_directory: Cow<'static, Path>) -> Self {
+        Self {
+            db_directory,
+            sync_mode: None,
+            reader_threads: Some(READER_THREADS_DEFAULT),
+            resize_algorithm: None,
+        }
+    }
```

Comment:
## `Config` 1/2
`cuprate_database`'s config now _requires_ a PATH.

`cuprate_blockchain` just passes the PATH we used before.